### ISSUE TITLE
Crash when an actor turn off his physical host

### ISSUE THREAD

**Issue by AdrienGougeon** [simgrid/simgrid#323]
I suppose there is a bug when an actor turn off its physical host.

In the [linked example](https://github.com/simgrid/simgrid/files/2901443/suicide_host.zip) there is a "host->turn_off();" and a "hosts[1]->turn_off();".

Those two lines turn off the same physical machine, one is done by the master on another host, while the other is done by an actor placed on the Host we want to turn off.

By commenting only one of the two you can see the differences:
 - if turned off by an actor from another host (here, the master) everything goes well.
 - if turned off by an actor placed on the physical host we want to turn off, simgrid crashes with the following message: 

Uncaught exception std::runtime_error: udd/agougeon/Documents/intership_m2/simgrid/src/simix/ActorImpl.cpp:209:yield: The Impossible Did Happen (yet again). Please report this bug.

To use the example, unzip it and then execute:
cmake .
make
./suicide_host

**Comment by mquinson** [comment]
This is a bit weird, because we have a test that is supposed to do exactly this. My guess is that we messed up the test and fail to kill the actor right at the time when it stops. Reading your MWE (thanks for that! that's precious!), I think that our test should read as follows:

```C++
  simgrid::s4u::Actor::create("sleep5_restarted", all_hosts[1], [&sleeper_done]() {
    assert_exit(1, 5);
    simgrid::s4u::this_actor::sleep_for(5);
    all_hosts[1]->turn_off(); // kill the host right at the end of this sleep and of this actor
    sleeper_done = true;
  });
  simgrid::s4u::this_actor::sleep_for(10);
  all_hosts[1]->turn_on();
  xbt_assert(sleeper_done, "Not sure of how the actor survived the shutdown of its host.");
```

Note that the actor is actually turning its host off itself, ensuring that it occurs exactly at the right time. Actually, I don't commit it because I confirm that this new test reproduces the pb that Adrien is experiencing, and I fail to understand that pb. 

@frs69wq @agiersch, any idea, maybe?

Thanks to all,
Mt.

**Comment by agiersch** [comment]
>   xbt_assert(sleeper_done, "Not sure of how the actor survived the shutdown of its host.");

Shouldn't it be "not sleeper_done" here?

Anyway, IIUC, the actor is inserted twice in actors_to_run. The test "actor != this" (ActorImpl.cpp:220) is not working when the actors suicides through a turn_off(). I don't know how to fix this properly ATM.

Arnaud

**Comment by mquinson** [comment]
Indeed you are right. But actually, the test is failing on a segfault, with the following backtrace:
```
#0  0x0000555555590160 in ?? ()
#1  0x00007ffff7c53af4 in std::_Function_handler<void(bool, void*), simgrid::s4u::Actor::on_exit(const std::function<void(bool)>&)::<lambda()>::<lambda(int, void*)> >::_M_invoke(const std::_Any_data &, bool &&, void *&&) (__functor=..., __args#0=@0x5555555902e0: 104, __args#1=@0x7ffff7cdd4b2: 0xc6ffffff58858b48)
    at /usr/include/c++/8/bits/std_function.h:299
#2  0x00007ffff7c53a96 in std::_Function_base::_Base_manager<simgrid::s4u::Actor::on_exit(const std::function<void(bool)>&)::<lambda()>::<lambda(int, void*)> >::_M_init_functor(std::_Any_data &, simgrid::s4u::Actor::<lambda()>::<lambda(int, void*)> &&) (__functor=..., __f=...) at /usr/include/c++/8/bits/std_function.h:220
#3  0x0000000000000000 in ?? ()
```
I'm not sure. All I can say is that I have a python example that is also failing if I add a at_exit in my code. But that's not the only reason, because this code fails even if I comment the line 131 of the current code: `assert_exit(true, 5);`

**Comment by agiersch** [comment]
I got different backtraces, e.g. with different context factories.  Since commit 26d1147ec, it should be consistent and abort early when using  `--log=simix_popping.thres:debug`.

I noticed another oddity with the tests from activity-lifecycle: if the main actor waits long enough between turn_off() and turn_on(), the test is failing with host containing more actors than expected.

The problem may occur with test_sleep_restart_end, but it's easier to show with test_sleep_restart_middle. Add for example `simgrid::s4u::this_actor::sleep_for(10);` between turn_off() and turn_on(). It's not failing if the sleep duration is less than 8, and it's failing when it's greater.
